[CI] SnapshotStressTestsIT testRandomActivities failing

Test failed because of a MasterNotDiscoveredException. Not sure if this was a random issue or worth investigating so raising the issue.

**Build scan:**
https://gradle-enterprise.elastic.co/s/6wiigan6zqcyy/tests/:server:internalClusterTest/org.elasticsearch.snapshots.SnapshotStressTestsIT/testRandomActivities

**Reproduction line:**
`./gradlew ':server:internalClusterTest' --tests "org.elasticsearch.snapshots.SnapshotStressTestsIT.testRandomActivities" -Dtests.seed=FF84BCB3D914AD7C -Dtests.locale=es-US -Dtests.timezone=America/Dawson -Druntime.java=11`

**Applicable branches:**
master

**Reproduces locally?:**
No

**Failure history:**
https://gradle-enterprise.elastic.co/scans/tests?tests.container=org.elasticsearch.snapshots.SnapshotStressTestsIT&tests.test=testRandomActivities

**Failure excerpt:**
```
org.elasticsearch.discovery.MasterNotDiscoveredException: (No message provided)

  at org.elasticsearch.action.support.master.TransportMasterNodeAction$AsyncSingleAction$2.onTimeout(TransportMasterNodeAction.java:253)
  at org.elasticsearch.cluster.ClusterStateObserver$ContextPreservingListener.onTimeout(ClusterStateObserver.java:324)
  at org.elasticsearch.cluster.ClusterStateObserver$ObserverClusterStateListener.onTimeout(ClusterStateObserver.java:241)
  at org.elasticsearch.cluster.service.ClusterApplierService$NotifyTimeout.run(ClusterApplierService.java:621)
  at org.elasticsearch.common.util.concurrent.ThreadContext$ContextPreservingRunnable.run(ThreadContext.java:678)
  at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
  at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
  at java.lang.Thread.run(Thread.java:834)

```

another failure which looks related to timeouts, but it froze?
https://gradle-enterprise.elastic.co/s/ypyj2dojbtnr2/tests/:server:internalClusterTest/org.elasticsearch.snapshots.SnapshotStressTestsIT/testRandomActivities

branch 7.17 - It failed on macos (it was mentioned that on mac it is slower), 
`./gradlew ':server:internalClusterTest' --tests "org.elasticsearch.snapshots.SnapshotStressTestsIT.testRandomActivities" -Dtests.seed=EC90F69804185ACD -Dtests.locale=sk -Dtests.timezone=Europe/Simferopol -Druntime.java=8`

```
   Caused by:
        java.lang.AssertionError: unexpected
            at __randomizedtesting.SeedInfo.seed([EC90F69804185ACD]:0)
            at org.elasticsearch.snapshots.SnapshotStressTestsIT.logAndFailTest(SnapshotStressTestsIT.java:181)
            at org.elasticsearch.snapshots.SnapshotStressTestsIT.access$000(SnapshotStressTestsIT.java:83)
            at org.elasticsearch.snapshots.SnapshotStressTestsIT$2.onFailure(SnapshotStressTestsIT.java:175)

            Caused by:
            MasterNotDiscoveredException[NodeNotConnectedException[[node_s2][127.0.0.1:49214] Node not connected]]; nested: NodeNotConnectedException[[node_s2][127.0.0.1:49214] Node not connected];
                at org.elasticsearch.action.support.master.TransportMasterNodeAction$AsyncSingleAction$2.onTimeout(TransportMasterNodeAction.java:297)
                at org.elasticsearch.cluster.ClusterStateObserver$ContextPreservingListener.onTimeout(ClusterStateObserver.java:345)
                at org.elasticsearch.cluster.ClusterStateObserver$ObserverClusterStateListener.onTimeout(ClusterStateObserver.java:263)
                at org.elasticsearch.cluster.service.ClusterApplierService$NotifyTimeout.run(ClusterApplierService.java:660)
                at org.elasticsearch.common.util.concurrent.ThreadContext$ContextPreservingRunnable.run(ThreadContext.java:718)
                at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
                at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
                at java.lang.Thread.run(Thread.java:748)

                Caused by:
                NodeNotConnectedException[[node_s2][127.0.0.1:49214] Node not connected]
                    at org.elasticsearch.transport.ClusterConnectionManager.getConnection(ClusterConnectionManager.java:257)
                    at org.elasticsearch.test.transport.StubbableConnectionManager.getConnection(StubbableConnectionManager.java:70)
                    at org.elasticsearch.transport.TransportService.getConnection(TransportService.java:895)
                    at org.elasticsearch.transport.TransportService.sendRequest(TransportService.java:791)
                    at org.elasticsearch.transport.TransportService.sendRequest(TransportService.java:779)
                    at org.elasticsearch.action.support.master.TransportMasterNodeAction$AsyncSingleAction.doStart(TransportMasterNodeAction.java:228)
                    at org.elasticsearch.action.support.master.TransportMasterNodeAction.doExecute(TransportMasterNodeAction.java:158)
                    at org.elasticsearch.action.support.master.TransportMasterNodeAction.doExecute(TransportMasterNodeAction.java:52)
                    at org.elasticsearch.action.support.TransportAction$RequestFilterChain.proceed(TransportAction.java:179)
                    at org.elasticsearch.action.support.TransportAction.execute(TransportAction.java:154)
                    at org.elasticsearch.action.support.TransportAction.execute(TransportAction.java:82)
                    at org.elasticsearch.client.node.NodeClient.executeLocally(NodeClient.java:95)
                    at org.elasticsearch.client.node.NodeClient.doExecute(NodeClient.java:73)
                    at org.elasticsearch.client.support.AbstractClient.execute(AbstractClient.java:407)
                    at org.elasticsearch.client.support.AbstractClient$ClusterAdmin.execute(AbstractClient.java:708)
                    at org.elasticsearch.action.ActionRequestBuilder.execute(ActionRequestBuilder.java:59)
                    at org.elasticsearch.snapshots.SnapshotStressTestsIT$TrackedCluster.lambda$pollForSnapshotCompletion$35(SnapshotStressTestsIT.java:1082)
                    at org.elasticsearch.snapshots.SnapshotStressTestsIT$1.doRun(SnapshotStressTestsIT.java:152)

```

--- a/server/src/internalClusterTest/java/org/elasticsearch/snapshots/SnapshotStressTestsIT.java
+++ b/server/src/internalClusterTest/java/org/elasticsearch/snapshots/SnapshotStressTestsIT.java
@@ -15,6 +15,7 @@ import org.elasticsearch.ExceptionsHelper;
 import org.elasticsearch.action.ActionListener;
 import org.elasticsearch.action.ShardOperationFailedException;
 import org.elasticsearch.action.StepListener;
+import org.elasticsearch.action.admin.cluster.health.ClusterHealthRequestBuilder;
 import org.elasticsearch.action.admin.cluster.health.ClusterHealthResponse;
 import org.elasticsearch.action.admin.cluster.node.hotthreads.NodeHotThreads;
 import org.elasticsearch.action.admin.cluster.snapshots.create.CreateSnapshotRequestBuilder;
@@ -574,13 +575,8 @@ public class SnapshotStressTestsIT extends AbstractSnapshotIntegTestCase {
                             snapshotInfo.snapshotId().getName()
                         );
 
-                        client().admin()
-                            .cluster()
-                            .prepareHealth(indicesToRestore)
-                            .setWaitForEvents(Priority.LANGUID)
-                            .setWaitForGreenStatus()
+                        prepareClusterHealthRequest(indicesToRestore).setWaitForGreenStatus()
                             .setWaitForNoInitializingShards(true)
-                            .setWaitForNodes(Integer.toString(internalCluster().getNodeNames().length))
                             .execute(mustSucceed(clusterHealthResponse -> {
 
                                 logger.info(
@@ -874,13 +870,7 @@ public class SnapshotStressTestsIT extends AbstractSnapshotIntegTestCase {
                         snapshotName
                     );
 
-                    client().admin()
-                        .cluster()
-                        .prepareHealth(targetIndexNames.toArray(new String[0]))
-                        .setWaitForEvents(Priority.LANGUID)
-                        .setWaitForYellowStatus()
-                        .setWaitForNodes(Integer.toString(internalCluster().getNodeNames().length))
-                        .execute(ensureYellowStep);
+                    prepareClusterHealthRequest(targetIndexNames.toArray(String[]::new)).setWaitForYellowStatus().execute(ensureYellowStep);
 
                     ensureYellowStep.addListener(mustSucceed(clusterHealthResponse -> {
                         assertFalse("timed out waiting for yellow state of " + targetIndexNames, clusterHealthResponse.isTimedOut());
@@ -1341,13 +1331,7 @@ public class SnapshotStressTestsIT extends AbstractSnapshotIntegTestCase {
 
                             logger.info("--> waiting for yellow health of [{}] prior to indexing [{}] docs", indexName, docCount);
 
-                            client().admin()
-                                .cluster()
-                                .prepareHealth(indexName)
-                                .setWaitForEvents(Priority.LANGUID)
-                                .setWaitForYellowStatus()
-                                .setWaitForNodes(Integer.toString(internalCluster().getNodeNames().length))
-                                .execute(ensureYellowStep);
+                            prepareClusterHealthRequest(indexName).setWaitForYellowStatus().execute(ensureYellowStep);
 
                             final StepListener<BulkResponse> bulkStep = new StepListener<>();
 
@@ -1412,6 +1396,17 @@ public class SnapshotStressTestsIT extends AbstractSnapshotIntegTestCase {
 
         }
 
+    }
+
+    // Prepares a health request with twice the default (30s) timeout that waits for all cluster tasks to finish as well as all cluster
+    // nodes before returning
+    private static ClusterHealthRequestBuilder prepareClusterHealthRequest(String... targetIndexNames) {
+        return client().admin()
+            .cluster()
+            .prepareHealth(targetIndexNames)
+            .setTimeout(TimeValue.timeValueSeconds(60))
+            .setWaitForNodes(Integer.toString(internalCluster().getNodeNames().length))
+            .setWaitForEvents(Priority.LANGUID);
     }
 
     private static String stringFromSnapshotInfo(SnapshotInfo snapshotInfo) {

